diff --git a/lib/ssh.js b/lib/ssh.js
index f650413dcd9998a57af4f8fa56e6648b3faba6fa..efa6e8b55b588c55843490c62349bc2759a9b5a7 100644
--- a/lib/ssh.js
+++ b/lib/ssh.js
@@ -1,30 +1,42 @@
 var Client = require('ssh2').Client,
   http = require('http');
 
-module.exports = function(opt) {
+module.exports = function (opt) {
   var conn = new Client();
   var agent = new http.Agent();
 
-  agent.createConnection = function(options, fn) {
-    conn.once('ready', function() {
-      conn.exec('docker system dial-stdio', function(err, stream) {
-        if (err) {
-          conn.end();
-          agent.destroy();
-          return;
-        }
+  agent.createConnection = function (options, fn) {
+    try {
+      conn.once('ready', function () {
+        conn.exec('docker system dial-stdio', function (err, stream) {
+          if (err) {
+            handleError(err);
+          }
 
-        fn(null, stream);
-
-        stream.once('close', () => {
-          conn.end();
-          agent.destroy();
+          fn(null, stream);
+          stream.addListener('error', (err) => {
+            handleError(err);
+          });
+          stream.once('close', () => {
+            conn.end();
+            agent.destroy();
+          });
         });
-      });
-    }).connect(opt);
-
-    conn.once('end', () => agent.destroy());
+      }).on('error', (err) => {
+        handleError(err);
+      })
+        .connect(opt);
+      conn.once('end', () => agent.destroy());
+      return agent;
+    } catch (err) {
+      handleError(err);
+    }
   };
 
-  return agent;
+  function handleError(err) {
+    conn.end();
+    agent.destroy();
+    throw err;
+  }
+
 };